linker script: __data_rom_start alignment problem

LuoZhongYao opened the pull request:
The alignment of the data segment should be determined by the
data segment itself, rather than relying on the previous segment
fix #21554 

Signed-off-by: ZhongYao Luo <LuoZhongYao@gmail.com>

ioannisg reviewed (commented):
I wonder if the alignment enforcement could be done as in the .bss section (for consistency)

jukkar commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -343,7 +343,7 @@ SECTIONS
 
```
Typo in commit subject, s/scirpt/script/

stephanosio commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -343,7 +343,7 @@ SECTIONS
         } GROUP_LINK_IN(RAMABLE_REGION)
 
-    SECTION_DATA_PROLOGUE(_DATA_SECTION_NAME,,)
+    SECTION_DATA_PROLOGUE(_DATA_SECTION_NAME, ALIGN(4), )
```
This patch should not be necessary because `SECTION_DATA_PROLOGUE` already specifies `ALIGN_WITH_INPUT`, which "forces the section alignment to the maximum alignment of all its input sections."

https://github.com/zephyrproject-rtos/zephyr/blob/7d2586533d5f27bb7fbe50fb03cc50a4dfe30077/include/linker/linker-tool-gcc.h#L122-L123

This basically results the following, which is not a valid :
```
_DATA_SECTION_NAME ALIGN(4) : ALIGN_WITH_INPUT
```

https://sourceware.org/binutils/docs-2.24/ld/Forced-Output-Alignment.html#Forced-Output-Alignment
https://sourceware.org/ml/binutils/2013-06/msg00246.html

LuoZhongYao replied:
https://sourceware.org/binutils/docs/ld/Forced-Output-Alignment.html, 
The ALIGN_WITH_INPUT attribute forces the difference between VMA and LMA to remain constant throughout the output section. 
The following two examples:

```
$ git log
commit 7d2586533d5f27bb7fbe50fb03cc50a4dfe30077 (HEAD -> master, upstream/master)
Author: Wentong Wu <wentong.wu@intel.com>
Date:   Sat Jan 4 13:18:08 2020 +0800

    power: add DEVICE_PM_LOW_POWER_STATE for device power management
    
    When system going to sleep state, make peripherals go to state
    DEVICE_PM_LOW_POWER_STATE which needs less time than state
    DEVICE_PM_SUSPEND_STATE to save more power.
    
    Signed-off-by: Wentong Wu <wentong.wu@intel.com>


$ west build -p auto -b stm32_min_dev_blue samples/basic/blinky
$ ninja -C build
$ readelf -l build/zephyr/zephyr.elf

Elf file type is EXEC (Executable file)
Entry point 0x8000bd9
There are 4 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x002848 0x08002794 0x08002794 0x00008 0x00008 R   0x4
  LOAD           0x0000b4 0x08000000 0x08000000 0x02b40 0x02b40 RWE 0x4
  LOAD           0x002bf4 0x20000ee0 0x08002b40 0x000ec 0x000ec RW  0x4
  LOAD           0x002ce0 0x20000000 0x20000000 0x00000 0x00ee0 RW  0x8

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     text _TEXT_SECTION_NAME_2 .ARM.exidx sw_isr_table devconfig rodata 
   02     datas initlevel _k_mutex_area 
   03     bss noinit
```

```
$ git diff
diff --git a/samples/basic/blinky/src/main.c b/samples/basic/blinky/src/main.c
index d5e99bbd7e..80601e0ce8 100644
--- a/samples/basic/blinky/src/main.c
+++ b/samples/basic/blinky/src/main.c
@@ -14,11 +14,13 @@
 /* 1000 msec = 1 sec */
 #define SLEEP_TIME     1000
 
+char __noinit t = 0;
 void main(void)
 {
        u32_t cnt = 0;
        struct device *dev;
 
+       t = 1;
        dev = device_get_binding(LED_PORT);
        /* Set LED pin as output */
        gpio_pin_configure(dev, LED, GPIO_DIR_OUT);
$ ninja -C build
$ readelf -l build/zephyr/zephyr.elf

Elf file type is EXEC (Executable file)
Entry point 0x8000be5
There are 4 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x002854 0x080027a0 0x080027a0 0x00008 0x00008 R   0x4
  LOAD           0x0000b4 0x08000000 0x08000000 0x02b4c 0x02b4c RWE 0x4
  LOAD           0x002c00 0x20000ee4 0x08002b4f 0x000ec 0x000ec RW  0x4
  LOAD           0x002cf0 0x20000000 0x20000000 0x00000 0x00ee1 RW  0x8

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     text _TEXT_SECTION_NAME_2 .ARM.exidx sw_isr_table devconfig rodata 
   02     datas initlevel _k_mutex_area 
   03     bss noinit
```

This is the code I extracted from the latest repository.The difference between Example 1 and Example 2 is that there is one more byte in the noinit area in Example 2.
Note the differences in lines 4, 5, and 6 below:

```
1: Program Headers:
2:  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
3:  EXIDX          0x002848 0x08002794 0x08002794 0x00008 0x00008 R   0x4
4:  LOAD           0x0000b4 0x08000000 0x08000000 0x02b40 0x02b40 RWE 0x4
5:  LOAD           0x002bf4 0x20000ee0 0x08002b40 0x000ec 0x000ec RW  0x4
6:  LOAD           0x002ce0 0x20000000 0x20000000 0x00000 0x00ee0 RW  0x8
```
```
1: Program Headers:
2:  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
3:  EXIDX          0x002854 0x080027a0 0x080027a0 0x00008 0x00008 R   0x4
4:  LOAD           0x0000b4 0x08000000 0x08000000 0x02b4c 0x02b4c RWE 0x4
5:  LOAD           0x002c00 0x20000ee4 0x08002b4f 0x000ec 0x000ec RW  0x4
6: LOAD           0x002cf0 0x20000000 0x20000000 0x00000 0x00ee1 RW  0x8
```

Line 6 is bss and noinit, and the sizes are 0xEE0 and 0xEE1. Line 5 is datas. The VMA of datas is 0x20000ee0 and 0x20000ee4, which are aligned. However, the LMA address 0x08002b40 is aligned, but 0x08002b4f is non-aligned. In Example 2,
VMA (0x20000ee4)-VMA (0x20000ee1) = 3,
LMA (0x08002b4f)-LMA (0x08002b4c) = 3

My English is not good, maybe it is not well expressed.

stephanosio replied:
By default, all linker script alignments specified for VMA.

An unaligned LMA of the data segment is not necessarily a problem because the data section is loaded from the LMA region to the VMA region during init and all run-time access is done to the VMA region; so, there should be no problem as long as the LMA-to-VMA copy procedure supports unaligned source address.

https://github.com/zephyrproject-rtos/zephyr/blob/7d2586533d5f27bb7fbe50fb03cc50a4dfe30077/kernel/init.c#L165-L168

stephanosio replied:
@LuoZhongYao Are you actually seeing faults related to this?

LuoZhongYao replied:
Yes. When the LMA is not aligned, the burn image will report data verification failure. And it will also run away.

stephanosio replied:
> When the LMA is not aligned, the burn image will report data verification failure. And it will also run away.

What program are you using to download firmware to your board? I think that may just be the flashing utility's fault.

stephanosio replied:
I tested this on ATSAME70 (firmware downloaded using OpenOCD) and it works just fine.

```
volatile char __noinit rubbish;
volatile unsigned int data = 0xdeadbeef;

void main(void)
{
        printk("rubbish = %x\n", rubbish);
        printk("data = %x\n", data);
}
```

```
stephanos@DOMINUS:~/Dev/zephyr-public/zephyr/samples/hello_world/build_e70$ readelf -lS zephyr/zephyr.elf
There are 25 section headers, starting at offset 0x5b8e8:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] text              PROGBITS        00400000 000094 00015c 00 WAX  0   0  4
  [ 2] _TEXT_SECTION_NAM PROGBITS        0040015c 0001f0 001c4a 00  AX  0   0  4
  [ 3] sw_isr_table      PROGBITS        00401da8 001e3c 000238 00  WA  0   0  4
  [ 4] devconfig         PROGBITS        00401fe0 002074 00006c 00  WA  0   0  4
  [ 5] rodata            PROGBITS        0040204c 0020e0 000120 00   A  0   0  4
  [ 6] .ramfunc          PROGBITS        20400000 002298 000000 00   W  0   0  1
  [ 7] bss               NOBITS          20400000 002298 00013e 00  WA  0   0  8
  [ 8] noinit            NOBITS          20400140 002298 000d41 00  WA  0   0  8
  [ 9] datas             PROGBITS        20400e84 002200 000018 00  WA  0   0  4
  [10] initlevel         PROGBITS        20400e9c 002218 00006c 00  WA  0   0  4
  [11] _k_mutex_area     PROGBITS        20400f08 002284 000014 00  WA  0   0  4
  [12] .comment          PROGBITS        00000000 002298 000032 01  MS  0   0  1
  [13] .debug_aranges    PROGBITS        00000000 0022d0 000bf0 00      0   0  8
  [14] .debug_info       PROGBITS        00000000 002ec0 028b5a 00      0   0  1
  [15] .debug_abbrev     PROGBITS        00000000 02ba1a 007d3b 00      0   0  1
  [16] .debug_line       PROGBITS        00000000 033755 00ec37 00      0   0  1
  [17] .debug_frame      PROGBITS        00000000 04238c 0017b4 00      0   0  4
  [18] .debug_str        PROGBITS        00000000 043b40 004c51 01  MS  0   0  1
  [19] .debug_loc        PROGBITS        00000000 048791 00ac8f 00      0   0  1
  [20] .debug_ranges     PROGBITS        00000000 053420 002860 00      0   0  8
  [21] .ARM.attributes   ARM_ATTRIBUTES  00000000 055c80 000031 00      0   0  1
  [22] .symtab           SYMTAB          00000000 055cb4 003200 10     23 409  4
  [23] .strtab           STRTAB          00000000 058eb4 00292f 00      0   0  1
  [24] .shstrtab         STRTAB          00000000 05b7e3 000104 00      0   0  1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  y (purecode), p (processor specific)

Elf file type is EXEC (Executable file)
Entry point 0x400bd5
There are 3 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000094 0x00400000 0x00400000 0x0216c 0x0216c RWE 0x4
  LOAD           0x002200 0x20400e84 0x0040216f 0x00098 0x00098 RW  0x4
  LOAD           0x002298 0x20400000 0x20400000 0x00000 0x00e81 RW  0x8

 Section to Segment mapping:
  Segment Sections...
   00     text _TEXT_SECTION_NAME_2 sw_isr_table devconfig rodata
   01     datas initlevel _k_mutex_area
   02     bss noinit
```

```
*** Booting Zephyr OS build zephyr-v2.1.0-581-g4e135d76a3c2  ***
rubbish = ef
data = deadbeef
```

LuoZhongYao replied:
I am using Keil uVision V5.27.1.0. Target DLL: NULink\Nu_Link.dll V3.00. 
Nuvoton does not provide openOCD interface, so currently only Keil uVision can be used

stephanosio replied:
FWIW, also tested on nRF51 (Cortex-M0) and it works just fine.
```
stephanos@DOMINUS:~/Dev/zephyr-public/zephyr/samples/hello_world/build_nrf51$ readelf -lS zephyr/zephyr.elf
There are 26 section headers, starting at offset 0x5c6b8:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] text              PROGBITS        00000000 0000b4 0000a8 00 WAX  0   0  4
  [ 2] _TEXT_SECTION_NAM PROGBITS        000000a8 00015c 0023bc 00  AX  0   0  4
  [ 3] .ARM.exidx        ARM_EXIDX       00002464 002518 000008 00  AL  2   0  4
  [ 4] sw_isr_table      PROGBITS        0000246c 002520 0000d0 00  WA  0   0  4
  [ 5] devconfig         PROGBITS        0000253c 0025f0 000054 00  WA  0   0  4
  [ 6] rodata            PROGBITS        00002590 002644 000114 00   A  0   0  4
  [ 7] .ramfunc          PROGBITS        20000000 0027dc 000000 00   W  0   0  1
  [ 8] bss               NOBITS          20000000 0027e0 000152 00  WA  0   0  8
  [ 9] noinit            NOBITS          20000158 0027e0 000d01 00  WA  0   0  8
  [10] datas             PROGBITS        20000e5c 002758 00001c 00  WA  0   0  4
  [11] initlevel         PROGBITS        20000e78 002774 000054 00  WA  0   0  4
  [12] _k_mutex_area     PROGBITS        20000ecc 0027c8 000014 00  WA  0   0  4
  [13] .comment          PROGBITS        00000000 0027dc 000032 01  MS  0   0  1
  [14] .debug_aranges    PROGBITS        00000000 002810 000b50 00      0   0  8
  [15] .debug_info       PROGBITS        00000000 003360 025908 00      0   0  1
  [16] .debug_abbrev     PROGBITS        00000000 028c68 00775c 00      0   0  1
  [17] .debug_line       PROGBITS        00000000 0303c4 00e5da 00      0   0  1
  [18] .debug_frame      PROGBITS        00000000 03e9a0 001774 00      0   0  4
  [19] .debug_str        PROGBITS        00000000 040114 00584f 01  MS  0   0  1
  [20] .debug_loc        PROGBITS        00000000 045963 00d761 00      0   0  1
  [21] .debug_ranges     PROGBITS        00000000 0530c8 003798 00      0   0  8
  [22] .ARM.attributes   ARM_ATTRIBUTES  00000000 056860 00002f 00      0   0  1
  [23] .symtab           SYMTAB          00000000 056890 003380 10     24 419  4
  [24] .strtab           STRTAB          00000000 059c10 002998 00      0   0  1
  [25] .shstrtab         STRTAB          00000000 05c5a8 00010f 00      0   0  1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  y (purecode), p (processor specific)

Elf file type is EXEC (Executable file)
Entry point 0x9a1
There are 4 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x002518 0x00002464 0x00002464 0x00008 0x00008 R   0x4
  LOAD           0x0000b4 0x00000000 0x00000000 0x026a4 0x026a4 RWE 0x4
  LOAD           0x002758 0x20000e5c 0x000026a7 0x00084 0x00084 RW  0x4
  LOAD           0x0027e0 0x20000000 0x20000000 0x00000 0x00e59 RW  0x8

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx
   01     text _TEXT_SECTION_NAME_2 .ARM.exidx sw_isr_table devconfig rodata
   02     datas initlevel _k_mutex_area
   03     bss noinit
```

Firmware downloaded using `nrfjprog`.

stephanosio replied:
> I am using Keil uVision V5.27.1.0. Target DLL: NULink\Nu_Link.dll V3.00.
> Nuvoton does not provide openOCD interface, so currently only Keil uVision can be used

Maybe this could work?
https://github.com/OpenNuvoton/OpenOCD-Nuvoton

LuoZhongYao replied:
This may be a tool issue, is it possible to place the datas section in front of the bss, so that the sections with LMA are continuous,
This will ensure that both LMA and VMA are aligned, and at the same time ALIGN_WITH_INPUT

LuoZhongYao replied:
I tried this driver without success

stephanosio replied:
I think the correct approach in this case would be to contact Nuvoton to fix their `Nu_Link.dll`.

Unaligned LMA is not really a bug; it can even be said to be beneficial because it results in reduced firmware image size (though, the practical significance of this is somewhat questionable).

LuoZhongYao replied:
LMA unaligned does not necessarily reduce the firmware image, or even increase it. Consider the case of noinit misalignment and rodata alignment

LuoZhongYao replied:
> 
> 
> This may be a tool issue, is it possible to place the datas section in front of the bss, so that the sections with LMA are continuous,
> This will ensure that both LMA and VMA are aligned, and at the same time ALIGN_WITH_INPUT

Can you consider this suggestion to ensure that both LMA and VMA are aligned

stephanosio replied:
cc @andrewboie @ioannisg

LuoZhongYao committed:
Fix __data_rom_start alignment problem

The alignment of the data segment should be determined by the
data segment itself, rather than relying on the previous segment

Signed-off-by: ZhongYao Luo <LuoZhongYao@gmail.com>
